In [1]:
from statsmodels.stats.power import TTestIndPower

## Exercise 1: Calculating Required Sample Size

In [3]:
# Defining parameters 
effect_size = 0.3
alpha = 0.05
power = 0.8


In [4]:
# calculate sample size required to detect 3% increase
analysis = TTestIndPower()
sample_size = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size:.2f}")

Required sample size: 175.38


#### Conclusion: The sample set of 176 is needed to ensure our test is properly powered

## Exercise 2: Understanding the Relationship Between Effect Size and Sample Size

In [ ]:
#1. Conditions: effect_size = 0.2, alpha = 0.05, power = 0.8
effect_size = 0.2
alpha = 0.05
power = 0.8
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")


Required sample size: 393.41


In [ ]:
#2.  Conditions: effect_size = 0.4, alpha = 0.05, power = 0.8
effect_size = 0.4
alpha = 0.05
power = 0.8
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")


Required sample size: 99.08


In [ ]:
# 3. Conditions: effect_size = 0.5, alpha = 0.05, power = 0.8
effect_size = 0.5
alpha = 0.05
power = 0.8
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")


Required sample size: 63.77


Conclusion: With an increase in effect size a smaller sample size is required. It is easier to detect a big change (bigger effect size) therefore smaller sample size. is needed.
- effect size (0.2) -> 394
- effect size (0.4) -> 100
- effect size (0.5) -> 64

## Exercise 3: Exploring the Impact of Statistical Power

In [9]:
#1. Conditions: effect_size = 0.2, alpha = 0.05, power = 0.7
effect_size = 0.2
alpha = 0.05
power = 0.7
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")

Required sample size: 309.56


In [10]:
#2. Conditions: effect_size = 0.2, alpha = 0.05, power = 0.8
effect_size = 0.2
alpha = 0.05
power = 0.8
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")

Required sample size: 393.41


In [11]:
#3. Conditions: effect_size = 0.2, alpha = 0.05, power = 0.9
effect_size = 0.2
alpha = 0.05
power = 0.9
analysis2 = TTestIndPower()
sample_size2 = analysis.solve_power(effect_size=effect_size,alpha=alpha,power=power)
print(f"Required sample size: {sample_size2:.2f}")

Required sample size: 526.33


Conclusion: With an increase in statistical power the sample size required increases as well. It is important while designing A/B testing because higher power ensures we are more likely to detect a true differece between two versions if one ectually exists.  
- power (0.7) -> 310
- power (0.8) -> 394
- power (0.9) -> 527

## Exercise 4: Implementing Sequential Testing

1. The stopping criteria for sequential testing:p < 0.02. Check results once per week for four weeks. We will stop if p<0.02.
2. The test would run for four weeks and we would calculate  p value that compares A vs B. Once we meet the criteria (p<0.02) the test will stop  and determine a winner, otherwise keep running until the end. At the end of the week 4 if it never stopped we will make a final decision using p<0.05.
3. If at the end of week 3 Version B has p-value = 0.02 (right at the limit) we would conclude that version B is very likely better. We need to make sure that we have enough data and results make sense.

## Exercise 5: Applying Bayesian A/B Testing

I would assume that a new feature has 50% chance of improving user engagement (it might imrove but I am not sure).If after collecting data there is a 65% chance I now believe that it's more likely that the new feature helps but i am not completely sure yet. Therefore, I need to continue testing for longer time. If later the probability goes up until 90% we can be sure to release this feature.
If posterior probability was only 55% it would mean that this feture doesn't change much.

## Exercise 6: Implementing Adaptive Experimentation

After the first week I would allocate 60% of traffic to layout C  and 20% each to layout A and layout B. In the next weeks I would check results again. If C is doing great I would give it even more traffic (70-80%). If another layout starts doing better, I would adjust my traffic accordingly. After a few weeks if one layout keeps winning, i would use it for everyone.
- Challenges: 
1) Stop testing other layouts too soon -> need to keep at least 10% of users on each version
2) Results keep changing -> wait for 2-3 weeks for stable results
3) Wrong conclusions -> once the winner is decided, needs to run final A/B testing to confirm
